### DISCUSSION TITLE
Allow generic arguments to be used with argument delegate (CS0411)

### DISCUSSION THREAD

**Discussion by 3fbaea00** [dotnet/csharplang#7733] [General]
I'm trying to make a method that return `delegate* unmanaged<T, T1, ..., TOut>` and accept a method. I'm tired of writing every time for method `TOut Meth(T, T1, ...)` castes in `delegate* unmanaged<T, T1, ..., TOut>`, like:
```
static bool Meth(nint ptr, int size) => true;
void Main()
{
   delegate* unmanaged<nint, int, bool> methPtr = (delegate* unmanaged<nint, int, bool>)&Meth;
}
```

I trying make:
```
static delegate* unmanaged<T, TOut> ldftn<T, TOut>(Func<T, TOut> meth) => (delegate* unmanaged<T, TOut>)meth.Method.MethodHandle.Value;
static delegate* unmanaged<T, T2, TOut> ldftn<T, T2, TOut>(Func<T, T2, TOut> meth) => (delegate* unmanaged<T, T2, TOut>)meth.Method.MethodHandle.Value;
// ...

static bool Meth(nint ptr, int size) => true;

void Main()
{
   delegate* unmanaged<nint, int, bool> methPtr = ldftn(Meth); // CS0411
}
```
But I get error: The type arguments for method 'ldftn<T, T2, TOut>(Func<T, T2, TOut>)' cannot be inferred from the usage. Try specifying the type arguments explicitly.
I can fix this by defining a generic type when calling:
```
delegate* unmanaged<nint, int, bool> methPtr = ldftn<nint, int, bool>(Meth);
```
BUT IF I DEFINE WRONG GENERICS, THEN AN ERROR CS0407 WILL COME UP:
```
delegate* unmanaged<MyWrongType, int, bool> methPtr = ldftn<MyWrongType, int, bool>(Meth); // CS0407
```
THAT THE Meth METHOD HAS THE WRONG ARGUMENTS OR RETURN TYPE, THAT IS, THE COMPILER KNOWS WHAT ARGUMENTS METH HAS AND IT WOULD BE LOGICAL THAT I DIDN’T WRITE THE GENERICS MYSELF.

**Comment by huoyaoyuan** [comment]
Not related to your question, but your code won't work because:

The result of `ldftn` isn't `MethodHandle.Value`. It's `MethodHandle.GetFunctionPointer()`.
Managed method (without `[UnmanagedCallersOnly]`) should not be passed into an unmanaged function pointer.

In addition, why don't you use the `&` operator to directly achieve function pointer from the method?

**Reply by 3fbaea00** [reply]
Thanks, but tbh, this is just show code. In real case I use post build events to change all calls of ldftn to ldftn opcode for performance.
About &, this returns `method group*` which can't be used directly, to get a real void* you need to convert `&Meth` to `(void*)(delegate* unmanaged<T, T2,...,TOut>)`
